Making the iron-ajax bubbling behavior configurable.

jpodwys opened the pull request:
I would like the ability to have:

* Flux-like ajax response handling and
* App-wide iron-ajax interceptors

To facilitate these, this PR proposes making iron-ajax event bubbling configurable, though default off. I'm willing to do more coding, but wanted to submit this PR to see if this is something the team is willing to consider before investing more time.

jpodwys commented:
I signed it!

rictic commented on iron-ajax.html:
```diff
@@ -299,6 +299,14 @@
       },
 
+      /**
+       * Whether the request, response, and error events should bubble to document.
```
document -> window

This would normally be super pedantic, but for the error event it's actually relevant.

rictic commented:
This seems reasonable to me, and it should have a low complexity and runtime cost.

The biggest cost I think is in adding another property to understand and more documentation to read. A clarifying sentence about why the events don't bubble by default should make that a bit easier to understand, like:

    These events do not bubble by default because the `error` event has special meaning on the window object.

jpodwys commented:
Thanks for the quick response yesterday! Based on your feedback, I updated my comment above the `bubbles` attribute, and added two unit tests. Please let me know what you think.

rictic commented on test/iron-ajax.html:
```diff
@@ -818,6 +824,39 @@
+          var total = 0;
+          function incrementTotal(){
+            total++;
+            if(total === 1) done();
```
why stop at 1?

rictic replied:
nits: 
* project style is to always use curly braces with if statements, even if they're just one line
* prefer a space between `if` and the opening paren

jpodwys replied:
Good catch! I had set it to `1` while testing and forgot to update it to `2`. Just updated it to `2` and tests are still passing. Also fixed the style issues.

rictic commented on test/iron-ajax.html:
```diff
@@ -818,6 +824,43 @@
+            }
+          }
+          window.addEventListener('request', incrementTotal);
+          document.addEventListener('response', incrementTotal);
```
Any reason for document here rather than window?

rictic commented:
Ok, that should be the final comment. If you change document to window, and the tests all pass this LGTM and will get merged :)

jpodwys commented:
Got it in, thanks for the code review!

rictic commented:
LGTM contingent on tests passing. Fixes #187.

rictic commented:
Thanks for the PR!

jpodwys commented:
Odd, I'm unable to reproduce the failures that travis encountered. And, AFAIK, I didn't change any code that affects the broken functionality--all the tests I added are passing even on travis.

Here's the test that failed: `when making a POST over the wire » FormData is handled correctly`. Any ideas?

rictic commented:
Hm, it may be flakey, rerunning.

jpodwys commented:
Looks like every single test that failed (both times) has to do with doing a "post over the wire" to `http://httpbin.org/post`. Could there be something wrong with that endpoint?

rictic commented:
Quite possibly. I'd added httpbin functionality to web-component-tester, just hadn't bothered to hook it up to these tests. Fixed with #190

Would you mind one last change to this PR to hit `/httpbin/post` rather than `http://httpbin.org/get`, or using the sinon fake server? I hadn't implemented `/get` in wct yet, as it only has the features that we've needed to test this element so far.

jpodwys commented:
Just to confirm, you'd like me to change

`<iron-ajax url="http://httpbin.org/get" bubbles></iron-ajax>`

to

`<iron-ajax url="/httpbin/post" bubbles></iron-ajax>`

Did you mean that I should change `get` to `post`? (The tests do not pass locally regardless of which I do. Will they be resolved with a rebase?)

jpodwys commented:
The reason I was having to hit a live endpoint for the first test (which confirms both `request` and `response`) are fired, is because sinon seems to swallow the `response` event. It's the same when I use `/httpbin/post`. Do you have another work-around?

rictic commented:
Yeah, it just merged in. Try rebasing, and changing

```html
<iron-ajax url="http://httpbin.org/get" bubbles></iron-ajax>
```

To:

```html
<iron-ajax url="/httpbin/post" method="POST" bubbles></iron-ajax>
```

jpodwys commented:
Moving this to #191
